homeassistant/components/emulated_kasa/__init__.py

```diff
@@ -0,0 +1,104 @@
+"""Support for local power state reporting of entities by emulating TP-Link Kasa smart plugs."""
+import logging
+from time import time
+
+from sense_energy import PlugInstance, SenseLink
+import voluptuous as vol
+
+import homeassistant.components as comps
+from homeassistant.const import (
+    ATTR_FRIENDLY_NAME,
+    CONF_ENTITIES,
+    CONF_NAME,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.template import Template
+
+from .const import CONF_POWER, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_ENTITY_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Required(CONF_POWER): vol.Any(
+            vol.Coerce(float),
+            cv.template,
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_ENTITIES): vol.Schema(
+                    {cv.entity_id: CONFIG_ENTITY_SCHEMA}
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Sense component."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+    hass.data[DOMAIN][CONF_ENTITIES] = conf.get(CONF_ENTITIES, {})
+
+    def devices():
+        """Drvices to be emulated."""
+        yield from get_plug_devices(hass)
+
+    server = SenseLink(devices)
+
+    async def stop_emulated_kasa(event):
+        await server.stop()
+
+    try:
+        await server.start()
+    except OSError as error:
+        _LOGGER.error("Failed to create UDP server at port 9999: %s", error)
+    else:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_emulated_kasa)
+
+    return True
+
+
+def get_plug_devices(hass):
+    """Produce list of plug devices from config entities."""
+    entity_ids = hass.data[DOMAIN][CONF_ENTITIES]
+    for entity_id in entity_ids:
+        state = hass.states.get(entity_id)
+        if state is None:
+            continue
+        name = state.attributes.get(ATTR_FRIENDLY_NAME, entity_id)
+        name = entities[entity_id].get(CONF_NAME, name)
+
```

Comment:
Power can come from a sensor like you have it now, or as switch provides the `ATTR_CURRENT_POWER_W` attribute

We can use `async_track_state_change_event` to watch it and update when it changes.

`from homeassistant.helpers.event import async_track_state_change_event`

`switch/__init__.py:ATTR_CURRENT_POWER_W = "current_power_w"`

hass.states.get(entity_id).attributes[ATTR_CURRENT_POWER_W]

Reply:
`HomeKit` and `min_max` (a more simple example) both use `async_track_state_change_event` to track entities and update.

Reply:
We also recently added `async_track_template_result` to `homeassistant.helpers.events` which would probably be more helpful for tracking power changes.  (git grep for a few usage examples)

I'd see myself using this for my UPS and my pool pump which probably cause a lot of confusion for sense
![Uploading Screen Shot 2020-09-03 at 7.00.54 PM.png…]()

Reply:
It might makes more sense to only allow `async_track_state_change_event` and just track specific entity state as users can always use the template integration to generate a a sensor to feed in the data.

Reply:
This integration doesn't broadcast the message, it replies to broadcasts:
  Sense broadcasts request for updates
  emulated_kasa replies with current power for devices

Since it only sends messages when requested by an outside device and may send repeated identical messages, I don't think it needs to track state changes

Reply:
So sense or the client is polling Home Assistant in this case.  Makes sense.

👍